src/fetch.js

```diff
@@ -136,13 +152,133 @@ function normalizeUrl(input) {
+  let current = concurrentRequests[cacheKey];
+  if (!current) {
+    current = next();
+    concurrentRequests[cacheKey] = current;
+    setTimeout(() => {
+      if (concurrentRequests[cacheKey] === current) {
+        delete concurrentRequests[cacheKey];
+      }
+    }, 1000).unref();
+    return current;
+  }
+  return current;
+}
+
 /**
  * Perform an HTTP Request
  * @param {Request} req
+ * @param {number} redirected
  * @returns {Response}
  */
 async function httpRequest(req, redirected = 0) {
-  let { protocol, host, port, hostname, pathname } = req.meta;
+  // For non GET requests, skip all connection pooling and caching logic.
+  if (req.method !== "GET") {
+    return realHttpRequest(req, redirected);
+  }
+
+  const { protocol, host, port, pathname } = req.meta;
+  const cacheKey = `${fetch.cacheBase}/${protocol}_${host}_${port}${pathname.replace(/_/g, '__').replace(/\//g, '_')}`;
+
+  // Combine concurrent requests for the same resource.
+  return checkConcurrent(cacheKey, async () => {
+
+    const config = cacheConfig[req.cache];
+    const { useCache, allowStale, conditionalRequest, updateCache, skipNet } = config;
+
+    const metaPath = `${cacheKey}.json`;
+    const bodyPath = `${cacheKey}`;
+
+    if (useCache) {
+      const metaBody = await readFileStream(metaPath).catch(err => {
+        if (!/^ENOENT:/.test(err.message)) throw err;
+      });
+      if (metaBody) {
+        const meta = JSON.parse(binToStr(await consume(metaBody)));
+        let fresh = Boolean(allowStale);
+
+        if (!allowStale) {
+          const now = Date.now();
+          const age = metaBody.stat.mtime - now;
+          // console.log("TODO: is stale?", metaPath, age, meta);
+          fresh = false;
+        }
+
+        if (!fresh && conditionalRequest) {
+          let condition = {};
+          const headers = new Headers(meta.headers);
+          const etag = headers.get('ETag');
+          const date = headers.get('Last-Modified') || headers.get('Date');
+          if (etag || date) {
+            if (date) req.headers.set('If-Modified-Since', date);
+            if (etag) req.headers.set('If-None-Match', etag);
+            const res = await realHttpRequest(req);
+            fresh = res.status === 304;
+          }
+        }
+
+        if (fresh) {
+          const body = await readFileStream(bodyPath);
+          const res = new Response(body, meta);
+          res.headers.set('Via', '1.1 fetch-cache');
+          if (body) res.cacheFile = Promise.resolve(bodyPath);
+          return res;
+        }
+
+
+      }
+    }
+
+
+    // If we're supposed to skip the net, we need to bail here.
+    if (skipNet) {
+      return new Response(undefined, {
+        url: req.url,
+        status: 504
+      });
+    }
+
+    // Make a real http request.
+    const res = await realHttpRequest(req, redirected);
+
+    if (updateCache && res.status === 200 && /^http/.test(res.url)) {
+      // TODO: make this cache update atomic!
```

Comment:
how likely is this to present itself in real world use cases?

Reply:
It's hard to say and we're fairly safe with the 1000ms timeout on concurrent requests, but there can be edge cases that don't play well with this.

For example, suppose it takes slightly longer than 1000ms to fetch and download a file and right after the concurrent timer expires another request for the same comes in.  It will try to serve up the partially written cache file thinking it's fully done. (though currently this is even less likely since we always assume cache to be stale and will first perform a conditional request before loading the file from disk).

This is one example of the many possible edge cases.  I don't think any are likely, but we should eventually make this more robust along with improving other aspects.